File: .net/Harvbot.Arms.Driver/HarvbotArmCircleNode.cs
using System;

namespace Harvbot.Arms.Driver
{
    /// <summary>
    /// Represents arm servo node.
    /// </summary>
    public class HarvbotArmCircleNode : HarvbotArmNode
    {
        /// <summary>
        /// Stores current node position.
        /// </summary>
        private float? pos;

        /// <summary>
        /// Represents Harvbot Arm node.
        /// </summary>
        /// <param name="type">The node type.</param>
        /// <param name="arm">The owner.</param>
        public HarvbotArmCircleNode(HarvbotArmNodeIdentifiers type, HarvbotArmBase arm) : base(type, arm)
        {
            this.pos = null;
        }

        /// <summary>
        /// Gets node type.
        /// </summary>
        public override HarvbotArmNodeTypes NodeType
        {
            get
            {
                return HarvbotArmNodeTypes.Circle;
            }
        }

        /// <summary>
        /// Moves node to specified position.
        /// </summary>
        /// <param name="degree">The degree.</param>
        public float? Move(float degree)
        {
            var response = this.SendCommand(HarvbotArmCommands.Move, degree.ToString());

            if (string.IsNullOrEmpty(response))
            {
                throw new InvalidOperationException($"Invalid moving: {degree}");
            }

            this.pos = float.Parse(response);

            return this.pos;
        }

        /// <summary>
        /// Gets current node position.
        /// </summary>
        /// <returns>The current node position degree.</returns>
        public float? GetPosition()
        {
            if (!this.pos.HasValue)
            {
                var response = this.SendCommand(HarvbotArmCommands.Position);

                if (!string.IsNullOrEmpty(response))
                {
                    this.pos = float.Parse(response);
                }
            }

            return this.pos;
        }
    }
}

File: linux/HarvbotArm/Nodes/HarvbotArmCircleNode.cpp
#include "HarvbotArmNode.h"
#include "HarvbotArmCircleNode.h"
#include "../HarvbotArmConstants.h"

HarvbotArmCircleNode::HarvbotArmCircleNode(HarvbotArmNodeIdentifiers identifier, float pos, float minPos, float maxPos) 
	: HarvbotArmNode(identifier)
{
	m_pos = pos;
	m_minPos = minPos;
	m_maxPos = maxPos;
}

HarvbotArmCircleNode::~HarvbotArmCircleNode(){
}

HarvbotArmNodeTypes HarvbotArmCircleNode::getType()
{
	return Circle;
}

float HarvbotArmCircleNode::currentPosition() {
	return m_pos;
}

float HarvbotArmCircleNode::setCurrentPosition(float pos)
{
	if (m_minPos > pos)
	{
		pos = m_minPos;
	}

	if (m_maxPos < pos)
	{
		pos = m_maxPos;
	}

	m_pos = pos;
}

float HarvbotArmCircleNode::move(float pos) 
{
	setCurrentPosition(pos);
}